packages/node/src/retryHandler.ts

```diff
@@ -0,0 +1,186 @@
+      return null;
+    }
+
+    return deviceToBufferMap.get(deviceId) || null;
+  }
+
+  // cleans up the id to buffer map if the job is done
+  private _cleanUpBuffer(userId: string, deviceId: string): void {
+    const deviceToBufferMap = this._idToBuffer.get(userId);
+    if (!deviceToBufferMap) {
+      return;
+    }
+
+    const eventsToRetry = deviceToBufferMap.get(deviceId);
+    if (!eventsToRetry?.length) {
+      deviceToBufferMap.delete(deviceId);
+    }
+
+    if (deviceToBufferMap.size === 0) {
+      this._idToBuffer.delete(userId);
+    }
+  }
+
+  private _onEventsError(events: ReadonlyArray<Event>): void {
+    events.forEach((event: Event) => {
+      const { user_id: userId = '', device_id: deviceId = '' } = event;
+      if (userId || deviceId) {
+        let deviceToBufferMap = this._idToBuffer.get(userId);
+        if (!deviceToBufferMap) {
+          deviceToBufferMap = new Map<string, Array<Event>>();
+          this._idToBuffer.set(userId, deviceToBufferMap);
+        }
+
+        let retryBuffer = deviceToBufferMap.get(deviceId);
+        if (!retryBuffer) {
+          retryBuffer = [];
+          deviceToBufferMap.set(deviceId, retryBuffer);
+          // In the next event loop, start retrying these events
+          setImmediate(() => this._retryEventsOnLoop(userId, deviceId));
+        }
+
+        this._eventsInRetry++;
+        retryBuffer.push(event);
+      }
+    });
+  }
+
+  private async _retryEventsOnLoop(userId: string, deviceId: string): Promise<void> {
+    const eventsBuffer = this._getRetryBuffer(userId, deviceId);
+    if (!eventsBuffer?.length) {
+      this._cleanUpBuffer(userId, deviceId);
+      return;
+    }
+
+    const initialEventCount = eventsBuffer.length;
+
+    let numRetries = 0;
+    const maxRetries = this._options.maxRetries;
+
+    while (numRetries < maxRetries) {
+      try {
+        // Don't try any new events that came in, to prevent overwhelming the api servers
+        const eventsToRetry = eventsBuffer.slice(0, initialEventCount);
+        const response = await this._transport.sendPayload(this._getPayload(eventsToRetry));
+        if (response.status === Status.Success) {
+          // Clean up the events
+          eventsBuffer.splice(0, initialEventCount);
+          this._eventsInRetry -= initialEventCount;
+          // Successfully sent the events, stop trying
+          break;
+        } else {
+          throw new Error(response.status);
+        }
+      } catch {
+        // Go on to next retry loop
+        numRetries += 1;
+        // If we hit the retry limit
+        if (numRetries === maxRetries) {
+          // We've tried the events for the max number of tries.
+          // Remove them permanently
+          eventsBuffer.splice(0, initialEventCount);
+          this._eventsInRetry -= initialEventCount;
+        } else {
+          // Exponential backoff - sleep for 2^(failed tries) ms before trying again
+          await asyncSleep(1 << numRetries);
```

Comment:
I think this is starting far too low. If events are being rejected because the device is throttled they're probably still going to be rejected 2ms later. Maybe 100ms or 500ms would be a better starting point? Along with thinking about the reduced payload you mentioned.